File: PA122.java
//leetcode: 53. Maximum Subarray
//Time cOmplexity: O(n)
//Space CComplexity: O(1)
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums==null || nums.length==0) return 0;
        int max= nums[0];
        int lsum=nums[0];
        for(int i=1;i<nums.length;i++){
            lsum= Math.max(lsum+nums[i], nums[i]);
            max= Math.max(lsum,max);
        }
        return max;
    }
}

File: PA122.py
#Time Complexity - O(n)
#Space Complexity - O(1)
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        cm = nums[0]
        gm = nums[0]
        for i in range(1,len(nums)):
            cm = max(nums[i],cm + nums[i])
            if gm < cm:
                gm = cm
                
        return gm